Sources/OpenCombine/Publishers/Publishers.ReplaceEmpty.swift

```diff
@@ -0,0 +1,181 @@
+        ///
+        /// - SeeAlso: `subscribe(_:)`
+        /// - Parameters:
+        ///     - subscriber: The subscriber to attach to this `Publisher`.
+        ///                   once attached it can begin to receive values.
+        public func receive<Downstream: Subscriber>(subscriber: Downstream)
+            where Upstream.Failure == Downstream.Failure,
+                  Upstream.Output == Downstream.Input
+        {
+            let inner = Inner(downstream: subscriber, output: output)
+            upstream.subscribe(inner)
+        }
+    }
+}
+
+extension Publishers.ReplaceEmpty {
+
+    private final class Inner<Downstream: Subscriber>
+        : Subscriber,
+          Subscription,
+          CustomStringConvertible,
+          CustomReflectable,
+          CustomPlaygroundDisplayConvertible
+        where Upstream.Failure == Downstream.Failure,
+              Upstream.Output == Downstream.Input
+    {
+
+        typealias Input = Upstream.Output
+        typealias Failure = Upstream.Failure
+
+        private let output: Output
+        private let downstream: Downstream
+
+        private var receivedUpstream = true
+        private var lock = UnfairLock.allocate()
+        private var downstreamRequested = false
+        private var finishedWithoutUpstream = false
+
+        private var status = SubscriptionStatus.awaitingSubscription
+
+        fileprivate init(downstream: Downstream, output: Output) {
+            self.downstream = downstream
+            self.output = output
+        }
+
+        deinit {
+            lock.deallocate()
+        }
+
+        func receive(subscription: Subscription) {
+            lock.lock()
+            guard case .awaitingSubscription = status else {
+                lock.unlock()
+                subscription.cancel()
+                return
+            }
+            status = .subscribed(subscription)
+            lock.unlock()
+            downstream.receive(subscription: self)
+            subscription.request(.unlimited)
+        }
+
+        func receive(_ input: Upstream.Output) -> Subscribers.Demand {
+            lock.lock()
+            guard case .subscribed = status else {
+                lock.unlock()
+                return .none
+            }
+            receivedUpstream = false
+            lock.unlock()
+            return downstream.receive(input)
+        }
+
+        func receive(completion: Subscribers.Completion<Upstream.Failure>) {
+            lock.lock()
+            guard case .subscribed = status else {
+                lock.unlock()
+                return
+            }
+            status = .terminal
+            lock.unlock()
+            switch completion {
+            case .finished:
+                guard downstreamRequested else {
+                    finishedWithoutUpstream = true
+                    return
+                }
+                if receivedUpstream {
+                    _ = downstream.receive(output)
+                }
```

Comment:
Oops. Reading from and writing to `downstreamRequested`, `finishedWithoutUpstream` and `receivedUpstream` should be locked with `lock`.

Here's what I suggest:

```swift
func receive(completion: Subscribers.Completion<Upstream.Failure>) {
    lock.lock()
    guard case .subscribed = status else {
        lock.unlock()
        return
    }
    status = .terminal
    if receivedUpstream {
        lock.unlock()
        downstream.receive(completion: completion)
        return
    }
    switch completion {
    case .finished:
        lock.unlock()
        if downstreamRequested {
            lock.unlock()
            _ = downstream.receive(output)
            downstream.receive(completion: completion)
            return
        }
        finishedWithoutUpstream = true
        lock.unlock()
    case .failure:
        lock.unlock()
        downstream.receive(completion: completion)
    }
}
```